### DISCUSSION TITLE
I am having a hard time to compile for Web using Ubuntu

### DISCUSSION THREAD

**Discussion by celebisait** [raysan5/raylib#2262] [General]
I have been following [this](https://github.com/raysan5/raylib/wiki/Working-for-Web-(HTML5)).

I have a fresh Ubuntu 20.04 LTS installation.

First I see we need emscripten. I have been following [here](https://emscripten.org/docs/getting_started/downloads.html) to install.

(This is on an absolutely new Ubuntu installation on Google cloud)

```
$ git clone https://github.com/emscripten-core/emsdk.git
$ cd emsdk
$ ./emsdk install latest
$ ./emsdk activate latest
$ source ./emsdk_env.sh
```

I also add `source "/home/celebisait_gmail_com/emsdk/emsdk_env.sh"` to my `~/.bashrc`.

Everything seems well until here.

I jump [here](https://emscripten.org/docs/getting_started/Tutorial.html#tutorial) to verify if everything is installed properly.

Looks like

```
$ emcc hello_world.c -o hello.html
```

works on a basic `hello_world.c` program.

Okay, we continue to raylib installation.

```
$ git clone https://github.com/raysan5/raylib.git
```

I add these to my `.profile`:

```
EMSDK_PATH=/home/celebisait_gmail_com/emsdk
EMSCRIPTEN_PATH=${EMSDK_PATH}/upstream/emscripten
CLANG_PATH=${EMSDK_PATH}/upstream/bin
PYTHON_PATH=/usr/bin/python3
NODE_PATH=${EMSDK_PATH}/node/12.9.1_64bit/bin
PATH=$PATH:${EMSDK_PATH}:${EMSCRIPTEN_PATH}:${CLANG_PATH}:${NODE_PATH}:${PYTHON_PATH}
```

Then I go to `raylib/src` and I do:

```
celebi@raylib-playground:~/raylib/src$ make PLATFORM=PLATFORM_WEB -B
emcc -c rcore.c -Wall -D_DEFAULT_SOURCE -DPLATFORM_WEB -DGRAPHICS_API_OPENGL_ES2 -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing  -std=gnu99 -Os -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw
make: emcc: Command not found
make: *** [Makefile:602: rcore.o] Error 127
celebi@raylib-playground:~/raylib/src$ emcc
emcc: error: no input files
celebi@raylib-playground:~/raylib/src$
```

I get `emcc` not found even though I obviously have `emcc`:

```
$ emcc
emcc: error: no input files
```

Any help would be appreciated.

**Comment by raysan5** [comment]
@celebisait it seems it could be some problem with the paths. I think emsdk includes some script to properly configure the paths: `emsdk_env.sh`. Also note that there are several `emcc` files (`emcc`, `emcc.py`, `emcc.bat`), maybe the system is not selecting the expected one to run?

**Comment by celebisait** [comment]
Yes, absolutely. I am already `source ./emsdk_env.sh` and also adding that to my `~/.profile` file. Here is the output after I run that:

```
Adding directories to PATH:
PATH += /home/celebisait_gmail_com/emsdk
PATH += /home/celebisait_gmail_com/emsdk/upstream/emscripten
PATH += /home/celebisait_gmail_com/emsdk/node/14.15.5_64bit/bin

Setting environment variables:
PATH = /home/celebisait_gmail_com/emsdk:/home/celebisait_gmail_com/emsdk/upstream/emscripten:/home/celebisait_gmail_com/emsdk/node/14.15.5_64bit/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
EMSDK = /home/celebisait_gmail_com/emsdk
EM_CONFIG = /home/celebisait_gmail_com/emsdk/.emscripten
EMSDK_NODE = /home/celebisait_gmail_com/emsdk/node/14.15.5_64bit/bin/node
Setting environment variables:
PATH = /home/celebisait_gmail_com/emsdk:/home/celebisait_gmail_com/emsdk/upstream/emscripten:/home/celebisait_gmail_com/emsdk/node/14.15.5_64bit/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/bin/python3
```

And it looks like to me `emcc` is installed properly, looking to this:

```
celebisait_gmail_com@raylib-playground:~/raylib/src$ emcc --check
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.0 (8e1e305519e1027726a48861a1fec5662f7e18a2)
shared:INFO: (Emscripten: Running sanity checks)
celebisait_gmail_com@raylib-playground:~/raylib/src$ which emcc
/home/celebisait_gmail_com/emsdk/upstream/emscripten/emcc
celebisait_gmail_com@raylib-playground:~/raylib/src$ 
```

I don't have much experience with `make`, I wish we could use `bazel` here. 
1. How does `make` calling `emcc`? If I can understand this better maybe we can fix it.
2. Should I try using `CMake` section instead?
3. Can anybody confirm that they were able to compile for Web using an Ubuntu instance recently?

As a side note: I thought having a fresh Ubuntu installation on cloud and following the [installation docs](https://github.com/raysan5/raylib/wiki/Working-for-Web-(HTML5)) very closely and carefully will lead me to success here, however, sadly, this wasn't the case.

**Reply by jamesl-github** [reply]
total guess, but try replacing 
emcc 
with 
./emcc 

from here 
https://emscripten.org/docs/tools_reference/emcc.html 

Command line syntax
emcc [options] file...
(Note that you will need ./emcc if you want to run emcc from your current directory.)

**Comment by celebisait** [comment]
@jamesl-github where do I change `emcc`? The issue is that I only get the error when I use `make` in raylib:

```
celebi@raylib-playground:~/raylib/src$ make PLATFORM=PLATFORM_WEB -B
```

`emcc` itself looks like working fine by itself.

**Comment by chriscamacho** [comment]
you need the path to emcc to be in your PATH env var too (looks like you just have emsdk in there)

**Comment by celebisait** [comment]
@chriscamacho thank you for the suggestion, I am trying to understand how I can actually address your comment.

So running `emcc` on a random directory, actually triggers `emcc`:

```
celebisait_gmail_com@raylib-playground:~/raylib/src$ emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.0 (8e1e305519e1027726a48861a1fec5662f7e18a2)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

When I check where it does come from:

```
celebisait_gmail_com@raylib-playground:~/raylib/src$ which emcc
/home/celebisait_gmail_com/emsdk/upstream/emscripten/emcc
```

It is coming from `/home/celebisait_gmail_com/emsdk/upstream/emscripten/` directory, which is in my `PATH` already:

```
celebisait_gmail_com@raylib-playground:~/raylib/src$ echo $PATH
/home/celebisait_gmail_com/emsdk:/home/celebisait_gmail_com/emsdk/upstream/emscripten:/home/celebisait_gmail_com/emsdk/node/14.15.5_64bit/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/bin/python3
```

What exactly are you suggestion to add? Thank you.

**Comment by jamesl-github** [comment]
I was thinking you should try changing this 
emcc -c rcore.c -Wall -D_DEFAULT_SOURCE -DPLATFORM_WEB -DGRAPHICS_API_OPENGL_ES2 -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing  -std=gnu99 -Os -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw

to this 
./emcc -c rcore.c -Wall -D_DEFAULT_SOURCE -DPLATFORM_WEB -DGRAPHICS_API_OPENGL_ES2 -Wno-missing-braces -Werror=pointer-arith -fno-strict-aliasing  -std=gnu99 -Os -I. -Iexternal/glfw/include -Iexternal/glfw/deps/mingw

the only change being  
emcc to ./emcc

**Comment by gtrxAC** [comment]
Seems like `make` on Ubuntu has a bug, I had this same issue with raylib and another project completely unrelated to raylib. I used `make -e` and it should find `emcc` and everything else just like running them from the command line normally.
